components/api/src/main/java/com/hotels/styx/api/service/spi/AbstractRegistry.java

```diff
@@ -43,6 +47,24 @@
 
     private final Announcer<Registry.ChangeListener> announcer = Announcer.to(ChangeListener.class);
     private final AtomicReference<Iterable<T>> snapshot = new AtomicReference<>(emptyList());
+    private final Predicate<Collection<T>> resourceConstraint;
+
+    /**
+     * Constructs an abstract registry with no resource constraint.
```

Comment:
Should it be a "registry" constraint now when it is in `AbstractRegistry`?

Resources are specific to the file based providers, right?

Reply:
I will copy + paste my previous comment:
`AbstractRegistry` has a method `protected static <T extends Identifiable> Changes<T> changes(Iterable<T> newResources, Iterable<T> currentResources)`, so the term is already used. We can still rename them to something else though.
`registryConstraint` is not really informative though, that would be naming it after the object it lives in, rather than the objects it constrains.
If we rename "resources" throughout the code, I will rename the constraint accordingly.

Reply:
Well, our `Resource` interface has methods like `path()`, `url()`, `inputStream()` etc which the `Identifiable`s stored in the Abstract registry do not have. So the question still remains what are the resources stored in the AbstractRegistry?

Maybe we can just decide that an identifiable stored in the registry in this instance can be called a resource and then we are sorted?

Another issue I have with this class is that it takes a type parameter T even if in practice it will never be anything else than a BackendService.

Reply:
We can't change the name unless we have a good alternative, so maybe let's leave that for later.

As for the type parameter, there is no point coupling this class to another class unnecessarily. Its purpose is to define a "Registry", so nothing else should be present.

Reply:
Right. Well I will approve.